File: apps/android/app/src/main/java/com/litter/android/ui/PiCapabilityGates.kt
package com.litter.android.ui

import androidx.compose.runtime.Composable
import uniffi.codex_mobile_client.AlleycatAgentRuntimeKind

/**
 * Capability gates for the Pi runtime on Android.
 *
 * Mirrors the iOS `PiCapabilityGates` surface: server-side enforcement
 * (network, plans, voice) lives in shared Rust on the connection layer;
 * this object exists purely as a render-time gate so Android UI hides
 * surfaces that do not exist for `AlleycatAgentRuntimeKind.Pi` runtimes.
 *
 * Today Pi runs in-process with no realtime voice transport and no app
 * plans flow, so both surfaces are hidden whenever the active runtime
 * is `AlleycatAgentRuntimeKind.Pi`. All other runtimes keep their
 * existing surfaces.
 */
object PiCapabilityGates {
    /** Whether the realtime voice launcher is visible for [runtime]. */
    fun isVoiceVisible(runtime: AlleycatAgentRuntimeKind): Boolean =
        when (runtime) {
            AlleycatAgentRuntimeKind.PI -> false
            else -> true
        }

    /** Whether the app/plan surfaces are visible for [runtime]. */
    fun isPlansVisible(runtime: AlleycatAgentRuntimeKind): Boolean =
        when (runtime) {
            AlleycatAgentRuntimeKind.PI -> false
            else -> true
        }

    /**
     * Returns `true` when the supplied [agentRuntimeKind] identifies a
     * pi server. Matches the canonical id used by
     * `AgentRuntimeKind::Pi` (serde rename: `"pi"`). Null and empty
     * inputs default to non-pi so cold-start UI does not flicker.
     */
    fun isPiRuntime(agentRuntimeKind: String?): Boolean =
        agentRuntimeKind?.trim()?.lowercase() == "pi"

    /**
     * String-based parity check mirroring iOS
     * `PiCapabilityGates.showsVoice(for:)`. Used by Compose call
     * sites that already have the runtime id as a string via the
     * shared Rust snapshot (e.g. `AppThreadSnapshot.agentRuntimeKind`).
     */
    fun showsVoice(agentRuntimeKind: String?): Boolean =
        !isPiRuntime(agentRuntimeKind)

    /** String-based parity check mirroring iOS `PiCapabilityGates.showsPlans(for:)`. */
    fun showsPlans(agentRuntimeKind: String?): Boolean =
        !isPiRuntime(agentRuntimeKind)
}

/**
 * Convenience Composable wrapper that only emits [content] when the
 * current [runtime] permits the voice surface. Equivalent to checking
 * [PiCapabilityGates.isVoiceVisible] at the call site; provided so call
 * sites stay declarative.
 */
@Composable
fun PiVoiceGate(runtime: AlleycatAgentRuntimeKind, content: @Composable () -> Unit) {
    if (PiCapabilityGates.isVoiceVisible(runtime)) {
        content()
    }
}

/**
 * Convenience Composable wrapper that only emits [content] when the
 * current [runtime] permits the plans surface.
 */
@Composable
fun PiPlansGate(runtime: AlleycatAgentRuntimeKind, content: @Composable () -> Unit) {
    if (PiCapabilityGates.isPlansVisible(runtime)) {
        content()
    }
}

File: apps/ios/Sources/Litter/Views/PiCapabilityGates.swift
import SwiftUI

/// Centralized capability gates for the pi runtime.
///
/// Per `architecture.md`, voice/realtime and the planning UI are
/// out-of-scope for pi v1. UI surfaces consult these gates so they can
/// hide the unsupported affordances whenever the active server's
/// agent runtime is pi. Server-side flag enforcement lives in Rust
/// (`connection.rs`); these gates are presentation-only.
enum PiCapabilityGates {
    /// Returns `true` when the supplied `agentRuntimeKind` identifies a
    /// pi server. `kind` matches the canonical id used by
    /// `AgentRuntimeKind::Pi` (serde rename: `"pi"`).
    static func isPiRuntime(_ agentRuntimeKind: String?) -> Bool {
        agentRuntimeKind?.lowercased() == "pi"
    }

    /// Whether voice / realtime affordances should be visible for a
    /// server whose runtime is `agentRuntimeKind`. Returns `false` for
    /// pi (no native realtime surface today) and `true` otherwise.
    static func showsVoice(for agentRuntimeKind: String?) -> Bool {
        !isPiRuntime(agentRuntimeKind)
    }

    /// Whether the planning UI (plan card, plan list, etc.) should be
    /// visible for a server whose runtime is `agentRuntimeKind`. Returns
    /// `false` for pi (pi's planning model differs enough that we don't
    /// render it inside Litter's plan card for v1) and `true` otherwise.
    static func showsPlans(for agentRuntimeKind: String?) -> Bool {
        !isPiRuntime(agentRuntimeKind)
    }
}

/// Convenience view modifier that hides its content when the active
/// server runtime kind is pi and the gate for `feature` is `false`.
///
/// Usage:
/// ```
/// VoiceOrbButton().piCapabilityGate(.voice, agentRuntimeKind: kind)
/// PlanCard().piCapabilityGate(.plans, agentRuntimeKind: kind)
/// ```
extension View {
    func piCapabilityGate(
        _ feature: PiCapabilityFeature,
        agentRuntimeKind: String?
    ) -> some View {
        let visible: Bool = {
            switch feature {
            case .voice: return PiCapabilityGates.showsVoice(for: agentRuntimeKind)
            case .plans: return PiCapabilityGates.showsPlans(for: agentRuntimeKind)
            }
        }()
        return Group {
            if visible {
                self
            } else {
                EmptyView()
            }
        }
    }
}

/// Features the pi capability gate covers.
enum PiCapabilityFeature {
    case voice
    case plans
}
